diff --git a/modules/deploy_managed_ui/deploy_managed_ui.module b/modules/deploy_managed_ui/deploy_managed_ui.module
index 014b5b6..33b612a 100644
--- a/modules/deploy_managed_ui/deploy_managed_ui.module
+++ b/modules/deploy_managed_ui/deploy_managed_ui.module
@@ -25,6 +25,12 @@ function deploy_managed_ui_form_alter(&$form, &$form_state, $form_id) {
   if (!empty($form['#node_edit_form'])) {
     deploy_managed_ui_form_elements($form, 'deploy_managed_ui_node_form_submit', TRUE);
   }
+  
+  if (isset($form['#form_id'])) {
+  		if ($form['#form_id'] == 'file_entity_edit') {
+  			deploy_managed_ui_form_elements($form, 'deploy_managed_ui_file_form_submit');
+  		}
+  }
 
   // For the rest we only support edit forms following the Entity API convention.
   if (!empty($form['#entity_type'])) {
@@ -73,7 +79,7 @@ function deploy_managed_ui_form_elements(&$form, $submit_handler, $vertical_tabs
   );
 
   $form['deploy_managed_ui']['text'] = array(
-    '#markup' => t('Select the plans to include these chanages in.'),
+    '#markup' => t('Select the plans to include these changes in.'),
   );
 
   $form['deploy_managed_ui']['plans'] = array(
@@ -86,7 +92,12 @@ function deploy_managed_ui_form_elements(&$form, $submit_handler, $vertical_tabs
     $form['actions']['#weight'] = ($weight + 1);
   }
 
-  $form['#submit'][] = $submit_handler;
+  if ($submit_handler == 'deploy_managed_ui_file_form_submit') {
+  	array_unshift($form['actions']['submit']['#submit'],$submit_handler);
+  } else {
+  	$form['#submit'][] = $submit_handler;
+  }
+  
 }
 
 /**
@@ -102,12 +113,28 @@ function deploy_managed_ui_node_form_submit(&$form, &$form_state) {
 }
 
 /**
+ * Submit handler for file forms supporting managed UI.
+ */
+function deploy_managed_ui_file_form_submit(&$form, &$form_state) {
+  
+  $form_state['file']->___deploy_plans = array();
+  foreach ($form_state['values']['deploy_managed_ui']['plans'] as $plan => $checked) {
+    if ($checked) {
+      $form_state['file']->___deploy_plans[$plan] = $plan;
+    }
+  }
+
+}
+
+/**
  * Submit handler for entity forms supporting managed UI.
  */
 function deploy_managed_ui_entity_form_submit(&$form, &$form_state) {
+  
   $entity_type = $form['#entity_type'];
   $entity = $form_state[$entity_type];
   $form_state[$entity_type]->___deploy_plans = array();
+  
   foreach ($form_state['values']['deploy_managed_ui']['plans'] as $plan => $checked) {
     if ($checked) {
       $form_state[$entity_type]->___deploy_plans[$plan] = $plan;
